test/ret_web/controllers/api-internal/change_login_email_controller_test.exs

```diff
@@ -0,0 +1,74 @@
+defmodule RetWeb.ApiInternal.V1.ChangeLoginEmailControllerTest do
+  use RetWeb.ConnCase
+  import Ret.TestHelpers
+  alias Ret.Account
+
+  @dashboard_access_header "x-ret-dashboard-access-key"
+  @dashboard_access_key "test-key"
+
+  setup_all do
+    merge_module_config(:ret, RetWeb.Plugs.DashboardHeaderAuthorization, %{dashboard_access_key: @dashboard_access_key})
+
+    on_exit(fn ->
+      Ret.TestHelpers.merge_module_config(:ret, RetWeb.Plugs.DashboardHeaderAuthorization, %{dashboard_access_key: nil})
+    end)
+  end
+
+  test "new email addresses must be valid", %{conn: conn} do
+    Account.find_or_create_account_for_email("alice@reticulum.io")
+    assert Account.exists_for_email?("alice@reticulum.io")
+    assert %{status: 400} = post_change_email_for_login(conn, "not_an_email_address", "alice@reticulum.io")
+    refute Account.exists_for_email?("not_an_email_address")
+    assert Account.exists_for_email?("alice@reticulum.io")
+  end
+
+  test "email addresses validation only applies to new emails", %{conn: conn} do
+    Account.find_or_create_account_for_email("not_an_email_address")
+    assert Account.exists_for_email?("not_an_email_address")
+    assert %{status: 200} = post_change_email_for_login(conn, "alice@reticulum.io", "not_an_email_address")
+    assert Account.exists_for_email?("alice@reticulum.io")
+    refute Account.exists_for_email?("not_an_email_address")
+  end
+
+  test "account emails can be changed", %{conn: conn} do
+    refute Account.exists_for_email?("alice@reticulum.io")
+    Account.find_or_create_account_for_email("alice@reticulum.io")
+    assert Account.exists_for_email?("alice@reticulum.io")
+    refute Account.exists_for_email?("alicia@anotherdomain.com")
+    assert %{status: 200} = post_change_email_for_login(conn, "alicia@anotherdomain.com", "alice@reticulum.io")
+    refute Account.exists_for_email?("alice@reticulum.io")
+    assert Account.exists_for_email?("alicia@anotherdomain.com")
+  end
+
+  test "emails cannot be shared between multiple accounts", %{conn: conn} do
+    Account.find_or_create_account_for_email("alice@reticulum.io")
+    Account.find_or_create_account_for_email("bob@reticulum.io")
+    assert %{status: 409} = post_change_email_for_login(conn, "bob@reticulum.io", "alice@reticulum.io")
+  end
+
+  test "email changes are rejected if old_email is not associated with an account", %{conn: conn} do
+    assert %{status: 409} = post_change_email_for_login(conn, "bob@reticulum.io", "alice@reticulum.io")
+    Account.find_or_create_account_for_email("bob@reticulum.io")
+    assert %{status: 409} = post_change_email_for_login(conn, "bob@reticulum.io", "alice@reticulum.io")
```

Comment:
`exercise-assert-exercise-assert` is a testing antipattern

Reply:
I removed the last two lines, but don't understand the comment / antipattern. Please explain.

Reply:
Happily! Each unit test is comprised of three stages:
1. Given…
2. when…
3. then…

Or put in the imperative mood:
1. Arrange.
2. Act.
3. Assert.

A unit test is a formal experiment and, as with any experiment, there is an inverse relationship between the number of variables and the reliability of the findings. By limiting our test to these three discrete stages we design the most conclusive test possible. Moreover, tests are the most reliable form of documentation and an experiment in three parts is eminently more readable than one with repeat stages.

If I arrange, act, assert, act, and assert again it leaves the question, “What is it that I am actually testing?”

When I write a test in Elixir I like to highlight the flow by diving the _Given…_ from the _when…_ and _then…_ with a blank line. The `when…` is always in the first line after the blank line.

**E.g.**
```elixir
do_some_setup()
foo = get_a_foo_i_need()

assert {:ok, bar} = the_behavior_under_test(foo)
assert something_about_bar(bar)
assert something_about_foo(foo)
```

Reply:
Gotcha. Thanks!